crates/biome_js_analyze/src/lint/nursery/no_substr.rs

```diff
@@ -0,0 +1,138 @@
+};
+use biome_console::markup;
+use biome_js_factory::make;
+use biome_js_syntax::{JsCallExpression, JsSyntaxToken};
+use biome_rowan::{AstSeparatedList, BatchMutationExt, TextRange, TokenText};
+
+use crate::JsRuleAction;
+
+declare_rule! {
+    /// Enforce the use of `slice()` over `substr()` and `substring()`.
+    ///
+    /// `slice()` is preferred over `substr()` and `substring()` because it is a more popular option with clearer behavior,
+    ///  and it has a consistent counterpart in arrays.
+    ///
+    /// Note that `substr`, `substring` and `slice` are not identical when arguments are passed.
+    /// For detailed differences, refer to the MDN documentation:
+    /// - [The difference between substring() and substr()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#the_difference_between_substring_and_substr)
+    /// - [Differences between substring() and slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#differences_between_substring_and_slice)
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substr(start, length);
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substring(indexStart, indexEnd);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// foo.slice(beginIndex, endIndex);
+    /// ```
+    ///
+    pub NoSubstr {
+        version: "next",
+        name: "noSubstr",
+        language: "js",
+        recommended: false,
+        fix_kind: FixKind::Safe,
+    }
+}
+
+impl Rule for NoSubstr {
+    type Query = Ast<JsCallExpression>;
+    type State = NoSubstrState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let callee = node.callee().ok()?;
+        let expression = callee.as_js_static_member_expression()?;
+        let value_token = expression.member().ok()?.value_token().ok()?;
+        let string_function_name = value_token.text_trimmed();
+
+        if matches!(string_function_name, "substr" | "substring") {
+            let arguments = node.arguments().ok()?;
+            let args = arguments.args();
+
+            Some(NoSubstrState {
+                value_token,
+                has_arguments: !args.is_empty(),
+            })
+        } else {
+            None
+        }
+    }
+
+    fn diagnostic(_: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let diagnostic_message = markup! {
+            "Avoid using "{state.member_name().text()}" and consider using slice instead."
+        }
+        .to_owned();
+        let note_message = {
+            markup! {
+                "Use "<Emphasis>"."{state.member_name().text()}"()"</Emphasis>" instead of "<Emphasis>".slice()"</Emphasis>"."
+            }
+            .to_owned()
+        };
```

Comment:
```suggestion
        let note_message =
            markup! {
                "Use "<Emphasis>"."{state.member_name().text()}"()"</Emphasis>" instead of "<Emphasis>".slice()"</Emphasis>"."
            }
            .to_owned();
```